File: src/main/java/com/cloudogu/gitops/infrastructure/git/providers/GitProvider.java
package com.cloudogu.gitops.infrastructure.git.providers;

import com.cloudogu.gitops.config.Credentials;
import com.cloudogu.gitops.utils.Tuple;

import java.net.URI;

public interface GitProvider {

	int REPO_TARGET_SEGMENT_COUNT = 2;

	/**
	 * Splits a "namespace/repoName" repo target into its namespace and name segments. Shared by
	 * providers that address repositories via a flat "namespace/name" string.
	 */
	static Tuple<String, String> splitRepoTarget(String repoTarget) {
		String[] parts = repoTarget.split("/", REPO_TARGET_SEGMENT_COUNT);
		return new Tuple<>(parts[0], parts[1]);
	}

	default boolean createRepository(String repoTarget, String description) {
		return createRepository(repoTarget, description, true);
	}

	boolean createRepository(String repoTarget, String description, boolean initialize);

	void setRepositoryPermission(String repoTarget, String principal, AccessRole role, Scope scope);

	default String repoUrl(String repoTarget) {
		return repoUrl(repoTarget, RepoUrlScope.IN_CLUSTER);
	}

	String repoUrl(String repoTarget, RepoUrlScope scope);

	String repoPrefix();

	Credentials getCredentials();

	URI prometheusMetricsEndpoint();

	String getUrl();

	String getProtocol();

	String getHost();

	String getGitOpsUsername();
}

File: src/main/kotlin/com/cloudogu/gitops/infrastructure/git/providers/GitProvider.kt
package com.cloudogu.gitops.infrastructure.git.providers

import com.cloudogu.gitops.config.Credentials
import java.net.URI

interface GitProvider {

    fun createRepository(repoTarget: String, description: String): Boolean {
        return createRepository(repoTarget, description, true)
    }

    fun createRepository(repoTarget: String, description: String, initialize: Boolean): Boolean

    fun setRepositoryPermission(repoTarget: String, principal: String, role: AccessRole, scope: Scope)

    fun repoUrl(repoTarget: String?): String {
        return repoUrl(repoTarget, RepoUrlScope.IN_CLUSTER)
    }

    fun repoUrl(repoTarget: String?, scope: RepoUrlScope): String

    fun repoPrefix(): String

    fun getCredentials(): Credentials?

    fun prometheusMetricsEndpoint(): URI?

    fun getUrl(): String?

    fun getProtocol(): String?

    fun getHost(): String?

    fun getGitOpsUsername(): String?
}

enum class AccessRole {
    READ, WRITE, MAINTAIN, ADMIN, OWNER
}

enum class Scope {
    USER, GROUP
}

/**
 * IN_CLUSTER: URLs intended for workloads running inside the Kubernetes cluster
 *             (e.g., ArgoCD, Jobs, in-cluster automation).
 *
 * CLIENT    : URLs intended for interactive or CI clients performing push/clone operations,
 *             regardless of their location.
 *             If the application itself runs inside Kubernetes, the Service DNS is used;
 *             otherwise, NodePort (for internal installations) or externalBase (for external ones)
 *             is selected automatically.
 */
enum class RepoUrlScope {
    IN_CLUSTER,
    CLIENT
}
